### PR TITLE
8352140: UBSAN: fix the left shift of negative value in klass.hpp, array_layout_helper()

### PR DESCRIPTION
**Pull request by afshin-zafari**
The `array_layout_helper()` with `jint tag` as its first arg, is called with a `tag` whose sign-bit is always set and considered as negative. This negative value is UB in left-shift operation.  Changing the type to `juint`  fixes this.

Tests:
linux-x64-debug tier1 with UBSAN enabled.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8352140](https://bugs.openjdk.org/browse/JDK-8352140): UBSAN: fix the left shift of negative value in klass.hpp, array_layout_helper() (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24184/head:pull/24184` \
`$ git checkout pull/24184`

Update a local copy of the PR: \
`$ git checkout pull/24184` \
`$ git pull https://git.openjdk.org/jdk.git pull/24184/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24184`

View PR using the GUI difftool: \
`$ git pr show -t 24184`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24184.diff">https://git.openjdk.org/jdk/pull/24184.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24184#issuecomment-2747532264)
</details>

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
That will fix it, but I can't help wonder whether `tag` is mis-typed in that function. In general that code seems very confused about signed vs unsigned and j(u)int versus (u)int. ??

File: src/hotspot/share/oops/klass.hpp [hunk 1]
```diff
@@ -524,7 +524,7 @@ class Klass : public Metadata {
     return l2esz;
   }
   static jint array_layout_helper(jint tag, int hsize, BasicType etype, int log2_esize) {
-    return (tag        << _lh_array_tag_shift)
+    return ((juint)tag << _lh_array_tag_shift)
```
Comment (dean-long) [inline src/hotspot/share/oops/klass.hpp:527]: Doesn't this turn the type of the return expression to unsigned, causing an implicit conversion back to signed, which is implementation-defined?  I think that's the reason for the weird-looking reinterpret_cast<> in JAVA_INTEGER_OP.

Comment (dholmes-ora) [reply src/hotspot/share/oops/klass.hpp:527]: Note we seem to do this a lot e.g.

> ./share/oops/klass.cpp:  `int  tag   =  isobj ? _lh_array_tag_obj_value : _lh_array_tag_type_value;`

`_lh_array_tag_type_value` is unsigned. So this call sequence goes from unsigned -> signed -> unsigned -> signed

Comment (kimbarrett) [reply src/hotspot/share/oops/klass.hpp:527]: Note that the discussion that led to the "weird-looking cast" in
JAVA_INTEGER_OP significantly predates the standard committee's decision to
enshrine two's-complement integers in C++20.  If we were to have that
discussion today my opinion would be quite different from what it was at the
time of that discussion.

Comment (afshin-zafari) [reply src/hotspot/share/oops/klass.hpp:527]: For my own learning: 
When developers use left-shift for doubling a value, then a negative operand may changed to a positive since the sign-bit may change. For example in
```
    signed short int x = -32768;
    signed short int y = x << 1;
``` 
the value of `y` would be `0`. So, when the left-shift is used as an arithmetic op, both the sign and size of the result/operand should be carefully considered. And, this is not dependent on C++xx.
So, left-shift of negative value is UB, until the developer explicitly decides on the type of the operand or the result.

Comment (kimbarrett) [reply src/hotspot/share/oops/klass.hpp:527]:
```
    signed short int x = -32768;
    signed short int y = x << 1;
```

That does seem like an interestingly weird case. Unless I'm missing something,
there's no UB-overflow in that. The shift expression promotes `short x` to
`int x`, sign extending it. The `int`-typed shift is fine (since C++20, and
effectively so prior to that in non-constexpr-required contexts - see below).
And the implicit conversion to `short y` is implementation-defined (before
C++20, though gcc may warn (-Woverflow) or fine (since C++20).

gcc warns about x being negative in C++11 to C++17 modes
(-Wshift-negative-value enabled by default), but doesn't treat it as UB.
Before C++20 gcc errors (warns if -fpermissive) if it's in a
required-constexpr-context, even if -Wshift-negative-value is disabled.
That all seems consistent.

Comment (afshin-zafari) [reply src/hotspot/share/oops/klass.hpp:527]: I had to emphasize that the case shown in the example may happen at run-time where compiler has no chance to warn/avoid/address it. 
My concern is that developers should not rely on the compiler to check the validation of left-shift op. They should be aware of the `signed` <-> `unsigned` conversions during the left-shift.
To find invalid cases of left-shift, UBSAN instruments them with assertions to catch them at run-time. If the assertion raised, good we found the problem. However, if no assertion raised for some left-shift ops, it doesn't mean that they are valid.

Comment (kimbarrett) [reply src/hotspot/share/oops/klass.hpp:527]: Is there a way to tell ubsan that we care about detecting overflows, but we do not care about detecting
left shift of a negative value?  Not that I can find, but maybe I missed something.  `-fsanitize=shift-base`
looks like it would check for both overflow and (prior to C++20) negative base.  We could disable
shift-base checking and do our own overflow assertion.  (Which might want to be packaged up in a helper,
as discussed in https://github.com/openjdk/jdk/pull/24196.)

Comment (afshin-zafari) [reply src/hotspot/share/oops/klass.hpp:527]: For the case in this PR, the left-shift results in overflow, since the operand is either 0xFFFFFFFF or 0xFFFFFFFE. Should we have two versions of `left_shift()` and `left_shift_no_overflow()`.

Comment (kimbarrett) [reply src/hotspot/share/oops/klass.hpp:527]: Sorry, I don't see where these 0xFFFFFFFF and 0xFFFFFFFE values come into
play.  (Also note that a left shift of a 32bit 0xFFFFFFFF cannot overflow
without exceeding the maximum allowed shift quantity (31), which is a
different UB form overflow.)

I think the fundamental problem here is that these masks and masking
operations are all using signed int, but should be using unsigned int.
Part of the problem here is (now) inappropriate use of jint.  (It might have
been slightly more appropriate when the code was originally written, as
[u]int32_t might not have been a thing then.  I suspect the basic framework
here is *very* old.)

Looking through this area more globally, I think things would be better if the
basic layout_helper type was uint32_t, with all the layout_helper_xxx
functions and masks adjusted accordingly.  I think that eliminates the need
for a whole mess of casts, and avoids UB from integer overflow.

So rather than working around one specific failure case, I suggest doing some
deeper cleaning here.

Related, I also just noticed a bogus test for "overflow" in
layout_helper_boolean_diffbit. UB is required for the assert to fail, so the
assert can be optimized away.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Comment by afshin-zafari** [comment]
I made the minimum `int->uint32` change because the layout_helper constants are `public` in the `Kalss` class and they are used all around Hotspot code, particularly in architecture dependent codes. 
Changing them breaks the build at SYMBOL generation phase. 
The two most-significant bits of the layout are used for array types and `lh < 0` (or more precisely `lh < _lh_neutral_value`) is used in C++ and assembly language of arch-dep code. 
@kimbarrett , can we limit the change here to `array_layout_helper` or should we proceed to use `uint32_t` instead?

File: src/hotspot/share/oops/klass.hpp [hunk 2]
```diff
@@ -523,7 +523,7 @@ class Klass : public Metadata {
            "sanity. l2esz: 0x%x for lh: 0x%x", (uint)l2esz, (uint)lh);
     return l2esz;
   }
-  static jint array_layout_helper(jint tag, int hsize, BasicType etype, int log2_esize) {
+  static jint array_layout_helper(uint32_t tag, int hsize, BasicType etype, int log2_esize) {
```
Comment (dean-long) [inline src/hotspot/share/oops/klass.hpp:526]:
```suggestion
  static jint array_layout_helper(jint tag, int hsize, BasicType etype, int log2_esize) {
    tag &= right_n_bits(_lh_array_tag_bits);
```
I'd like to propose a simpler fix that doesn't involve changing to unsigned.  In my experiments this satisfies UBSAN while still generating the same optimized code in product builds.

**Comment by kimbarrett** [comment]
> I made the `int->uint32` change at the minimum level, because the layout_helper constants are `public` in the `Kalss` class and are used all around Hotspot code, particularly in architecture dependent codes. Changing them to unsigned breaks the build (at SYMBOL generation phase). In addition, the two most-significant bits of the layout are used for `array` types and `lh < 0` (or more precisely `lh < _lh_neutral_value`) is used in C++ and assembly language of arch-dep code. @kimbarrett , can we limit the change here to `array_layout_helper` or should we proceed to use `uint32_t` instead?

I'm inclined against uglifying the code for a ubsan issue just to keep the
change small. I'd rather we did the "right" fix instead, even though there's
some fannout. I don't know how urgent we consider fixing the ubsan issue
though.

But if we're going to do a minimal change, I like the @dean-long suggestion here:
https://github.com/openjdk/jdk/pull/24184/files#r2090835514

The latest proposal also doesn't do anything about the issue I mentioned at
the end of this comment:
https://github.com/openjdk/jdk/pull/24184#discussion_r2055739014
where there's a related problem with layout_helper_boolean_diffbit.
That could be treated as a separate bug, in which case just file it.

**Comment by afshin-zafari** [comment]
> > I made the `int->uint32` change at the minimum level, because the layout_helper constants are `public` in the `Kalss` class and are used all around Hotspot code, particularly in architecture dependent codes. Changing them to unsigned breaks the build (at SYMBOL generation phase). In addition, the two most-significant bits of the layout are used for `array` types and `lh < 0` (or more precisely `lh < _lh_neutral_value`) is used in C++ and assembly language of arch-dep code. @kimbarrett , can we limit the change here to `array_layout_helper` or should we proceed to use `uint32_t` instead?
> 
> I'm inclined against uglifying the code for a ubsan issue just to keep the change small. I'd rather we did the "right" fix instead, even though there's some fannout. I don't know how urgent we consider fixing the ubsan issue though.
> 
> But if we're going to do a minimal change, I like the @dean-long suggestion here: https://github.com/openjdk/jdk/pull/24184/files#r2090835514
> 
> The latest proposal also doesn't do anything about the issue I mentioned at the end of this comment: [#24184 (comment)](https://github.com/openjdk/jdk/pull/24184#discussion_r2055739014) where there's a related problem with layout_helper_boolean_diffbit. That could be treated as a separate bug, in which case just file it.


Thanks for your input.
OK. I also agree. Then since the required changes are not in my expertise, I withdraw this PR and let qualified developers fix the issue.

For the bogus code in `layout_helper_boolean_diffbit`, [this](https://bugs.openjdk.org/browse/JDK-8358957)  is filed.

**Comment by afshin-zafari** [comment]
Withdrawn
